gdb/riscv-tdep.c

```diff
@@ -1301,42 +1241,139 @@ riscv_gdbarch_init (struct gdbarch_info info,
   dwarf2_append_unwinders (gdbarch);
   frame_unwind_append_unwinder (gdbarch, &riscv_frame_unwind);
 
-  /* Check any target description for validity.  */
+  /* Register time!
+   * We support two ways of dealing with registers:
+   * 1. The server sends gdb an XML description of its registers. This is what
+   * gdb calls tdesc. This way the server can also let us know what registers
+   * actually exist on the target.
+   * 2. Communicate with the target using register numbers only.
+   */
+
+  // First, for our own records, build a structure with all relevant
+  // information about registers.
+  // riscv_reg_info statically gets info about all registers except for the
+  // CSRs. We add those programmatically because there are many of them.
+  struct {
+      const char *name;
+      unsigned num;
+  } named_csr[] = {
+#define DECLARE_CSR(name, num) {#name, RISCV_ ## num ## _REGNUM},
+#include "opcode/riscv-opc.h"
+#undef DECLARE_CSR
+  };
+
+  static bool reg_info_built = false;
+  if (!reg_info_built)
+    {
+      for (unsigned i = 0; i < sizeof(named_csr) / sizeof(*named_csr); i++)
+        {
+          char *generic_name = (char*) malloc(8);
```

Comment:
This fails if there are more than 9999 CSRs.  I would suggest checking for that and aborting to avoid writing past the end of the malloced area.

Reply:
Added an assert.